module/zfs/bpobj.c

```diff
@@ -780,6 +779,65 @@ bpobj_enqueue_subobj(bpobj_t *bpo, uint64_t subobj, dmu_tx_t *tx)
 
 }
 
+/*
+ * Prefetch metadata required for bpobj_enqueue_subobj().
+ */
+void
+bpobj_prefetch_subobj(bpobj_t *bpo, uint64_t subobj)
+{
+	dmu_object_info_t doi;
+	bpobj_t subbpo;
+	uint64_t subsubobjs;
+	boolean_t copy_subsub = B_TRUE;
+	boolean_t copy_bps = B_TRUE;
+
+	ASSERT(bpobj_is_open(bpo));
+	ASSERT(subobj != 0);
+
+	if (subobj == dmu_objset_pool(bpo->bpo_os)->dp_empty_bpobj)
+		return;
+
+	VERIFY3U(0, ==, bpobj_open(&subbpo, bpo->bpo_os, subobj));
```

Comment:
Sure. Makes sense. Done. Just what is a chance it won't crash on actual operation?

Reply:
It looks to me like there are a few places in the existing bpobj code where it would still crash.  Which isn't great, but we can at least not make it worse.  We have been systematically trying to improve this kind of this over the years.  Clearly there's more work to be done.